clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,751 @@
+{-|
+Copyright  :  (C) 2024, QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+-}
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
```

Comment:
Is there a great reason to use `AllowAmbiguousTypes`? It makes error messages really bad. (I know you said very new GHCs do this better, but we're not even close to dropping most of them messing up..)

Reply:
Newer GHCs do much better on this! Maybe you should consider to update :stuck_out_tongue_winking_eye:.

Reply:
The alternative would be to use proxies, which I used in my [finite](https://hackage.haskell.org/package/finite-1.4.1.2/docs/Finite.html) library in the past, but the usage experience turns out to be really unsatisfactory. Using ambiguous types instead definitely improves the situation here.

IMO users that like to stick with older GHCs and don't like the bad error messages just shouldn't use the new feature then. As it's just the error messages tough, there is no problem in developing with a recent GHC and supporting the old ones with the error-free code in the end. 

I developed this interface with GHC 9.10 and I definitely experience several ambiguous type errors, which all where easy to identify and fix, but I never had to inspect any of the error messages of older GHCs.

Reply:
I was unable to find an example/test module that uses these new definitions in the commit, but if the issue here is using `foo @3` vs `foo (Proxy @3)` etc., if you are aiming for GHC 9.10 you should consider using a required type argument (i.e. using `forall n -> ...` instead of `forall n. ...`) so you can write it as `foo 3`.

Reply:
@gergoerdi I haven't looked into how the interface would particularly change with required type arguments yet, but my experience is that you don't need to pass the type explicitly in most of the cases. 

Thus in terms of writing code, I'd say that ambiguous types are the most elegant solution here :sparkles:.